specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,712 @@
+- `attester_slashing` - This topic is used solely for propagating attester slashings to proposers on the network. Attester slashings are sent in their entirety. Clients who receive an attester slashing on this topic MUST validate the conditions within `process_attester_slashing` before forwarding it across the network.
+
+#### Interop
+
+Unaggregated attestations from all shards are sent to the `beacon_attestation` topic.
+
+#### Mainnet
+
+Shards are grouped into their own subnets (defined by a shard topic). The number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard `shard_number % SHARD_SUBNET_COUNT` is assigned to the topic: `shard{shard_number % SHARD_SUBNET_COUNT}_beacon_attestation`. Unaggregated attestations are sent to the subnet topic. Aggregated attestations are sent to the `beacon_attestation` topic.
+
+### Messages
+
+Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24) has a maximum size of `GOSSIP_MAX_SIZE`.
+
+Clients MUST reject (fail validation) messages that are over this size limit. Likewise, clients MUST NOT emit or propagate messages larger than this limit.
+
+The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:
+
+
+| Topic                        | Message Type      |
+|------------------------------|-------------------|
+| beacon_block                 | BeaconBlock       |
+| beacon_attestation           | Attestation       |
+| shard{N}\_beacon_attestation | Attestation       |
+| voluntary_exit               | VoluntaryExit     |
+| proposer_slashing            | ProposerSlashing  |
+| attester_slashing            | AttesterSlashing  |
+
+Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.
+
+### Encodings
+
+Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.
+
+#### Interop
+
+- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.
+
+#### Mainnet
+
+- `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.
+
+Implementations MUST use a single encoding. Changing an encoding will require coordination between participating implementations.
+
+## The discovery domain: discv5
```

Comment:
I know this has been discussed in the past but just to try to get a definitive answer: the current discv5 spec uses RLP and the current eth2 networking spec depends on SSZ. Should we simply choose 1 serialization format for all networking or keep both?

Reply:
Alternatively, the ENR spec can be extended to support multiple serialisation formats, e.g. by pretending a prefix that identifies the encoding.

Reply:
multiple serialization formats means multiple signatures, and having everyone that publishes enr records understand all formats anyway, if I'm not mistaken. no real win. 

what perhaps would make more sense would be to use protobuf or some other established extensible format for enr's given that the requirements are significantly different from blockchain data, but that's probably unpalatable for eth1. regardless, this seems off topic for this thread and belongs in the discv5 forum, as rlp has been deemed unworkable for eth2.

Reply:
Opened a discussion: https://github.com/ethereum/devp2p/issues/107, and added to the open points tracker comment here: https://github.com/ethereum/eth2.0-specs/pull/1328#issuecomment-517664569.